src/util/client/mod.rs

```diff
@@ -0,0 +1,145 @@
+use crate::core::WithTracing;
+use crate::openid::TokenProvider;
+use crate::{error::ClientError, openid::TokenInjector};
+
+use async_trait::async_trait;
+use reqwest::{Response, StatusCode};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::marker::Send;
+use url::Url;
+
+/// A drogue HTTP client, backed by reqwest.
+
+#[async_trait]
+pub trait Client<TP>
+where
+    TP: TokenProvider,
+{
+    /// Constructor
+    fn new(client: reqwest::Client, url: Url, token_provider: TP) -> Self;
```

Comment:
I find it a bit strange to have the constructor as part of the trait, as you normally know the type at construction time, and there is no need for restricting the constructor signature I think.

Reply:
yep I agree. It also force the user to import the trait as well, which is annoying. I removed it on the other PR :)